library/common/http/client.cc

```diff
@@ -0,0 +1,388 @@
+              data.length, end_stream);
+    direct_stream->request_decoder_->decodeData(*buf, end_stream);
+  }
+
+  return ENVOY_SUCCESS;
+}
+
+envoy_status_t Client::sendMetadata(envoy_stream_t, envoy_headers) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+envoy_status_t Client::sendTrailers(envoy_stream_t stream, envoy_headers trailers) {
+  ASSERT(dispatcher_.isThreadSafe());
+  Client::DirectStreamSharedPtr direct_stream = getStream(stream);
+  // If direct_stream is not found, it means the stream has already closed or been reset
+  // and the appropriate callback has been issued to the caller. There's nothing to do here
+  // except silently swallow this.
+  // TODO: handle potential race condition with cancellation or failure get a stream in the
+  // first place. Additionally it is possible to get a nullptr due to bogus envoy_stream_t
+  // from the caller.
+  // https://github.com/lyft/envoy-mobile/issues/301
+  if (direct_stream) {
+    RequestTrailerMapPtr internal_trailers = Utility::toRequestTrailers(trailers);
+    ENVOY_LOG(debug, "[S{}] request trailers for stream:\n{}", stream, *internal_trailers);
+    direct_stream->request_decoder_->decodeTrailers(std::move(internal_trailers));
+  }
+
+  return ENVOY_SUCCESS;
+}
+
+envoy_status_t Client::cancelStream(envoy_stream_t stream) {
+  ASSERT(dispatcher_.isThreadSafe());
+  Client::DirectStreamSharedPtr direct_stream = getStream(stream);
+  if (direct_stream) {
+    removeStream(direct_stream->stream_handle_);
+
+    // Testing hook.
+    synchronizer_.syncPoint("dispatch_on_cancel");
+    direct_stream->callbacks_->onCancel();
+
+    // Since https://github.com/envoyproxy/envoy/pull/13052, the connection manager expects that
+    // response code details are set on all possible paths for streams.
+    direct_stream->setResponseDetails(getCancelDetails());
+
+    // The runResetCallbacks call synchronously causes Envoy to defer delete the HCM's
+    // ActiveStream. We have some concern that this could potentially race a terminal callback
+    // scheduled on the same iteration of the event loop. If we see violations in the callback
+    // assertions checking stream presence, this is a likely potential culprit. However, it's
+    // plausible that upstream guards will protect us here, given that Envoy allows streams to be
+    // reset from a wide variety of contexts without apparent issue.
+    direct_stream->runResetCallbacks(StreamResetReason::RemoteReset);
+  }
+  return ENVOY_SUCCESS;
+}
+
+const HttpClientStats& Client::stats() const { return stats_; }
+
+Client::DirectStreamSharedPtr Client::getStream(envoy_stream_t stream) {
+  auto direct_stream_pair_it = streams_.find(stream);
+  // Returning will copy the shared_ptr and increase the ref count.
+  return (direct_stream_pair_it != streams_.end()) ? direct_stream_pair_it->second : nullptr;
+}
+
+void Client::removeStream(envoy_stream_t stream_handle) {
+  RELEASE_ASSERT(dispatcher_.isThreadSafe(),
```

Comment:
Should this be an ASSERT like above, or should those calls be RELEASE_ASSERTS like this?

Reply:
Well... pick your poison I guess. I made the ones above `ASSERT` because they're mostly to validate that this large change doesn't have unexpected side effects (and of course we run with regular `ASSERT`s enabled). I don't _expect_ it to have context-changing side effects, so I didn't make them `RELEASE_ASSERTS`. I think you could probably make a reasonable argument either way, and I don't have particularly strong feelings.

Like I mentioned above, I added these `ASSERT`s everywhere I removed a previously-inline `post`.

I guess my slight longer-term bias would be to `ASSERT`, since it has the effect of documenting how the class is expected to be used, but doesn't necessarily incur the overhead of checking all the time for something that really shouldn't be happening.